### PR TITLE
2564 nested comments

### PR DESCRIPTION
**Pull request by balhelhas**
PR Checklist

- [x] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

This PR adds the functionality to reply to comments in the Researches and How-tos.

## Git Issues

Closes #[2564](https://github.com/ONEARMY/community-platform/issues/2564)

## Screenshots/Videos

2583FCC9-F50E-4C2D-A6F3-1E916C008B16

## What happens next?

Thanks for the contribution! We try to make sure all PRs are reviewed ahead of a monthly dev call (first Monday of the month, open to all!).

If the PR is working as intended it'll be merged and included in the next platform release, if not changes will be requested and re-reviewed once updated.

If you need more immediate feedback you can try reaching out on Discord in the [Community Platform `development` channel](https://discord.com/channels/586676777334865928/938781727017558018).

### CODE REVIEW

**Comment by balhelhas** [comment]
Hi @davehakkens, this is not finished but I opened the draft so you can check if the behaviours are of the "n replies" and "reply" button are correct. Currently is only available in the Researches module. Also, I still need to check with the team if they are ok with my approach here.

File: src/models/comment.model.ts [hunk 1]
```diff
@@ -14,4 +14,5 @@ export interface IComment {
   creatorCountry?: string | null
   text: string
   isUserVerified?: boolean
+  replies?: IComment[]
```
Comment (thisislawatts) [inline src/models/comment.model.ts:17]: **question:**  Have you considered alternative to nested data for storing this information, such as nested sub collections or root level collections.

As documented in the following:
https://firebase.google.com/docs/firestore/manage-data/structure-data

**Comment by davehakkens** [comment]
nice. Behaviour works for me! Can imagine some front-end edits might come later once we start using it. But can be released like this since it will be for `beta-tester` only in the beginning. 

(make sure to wrap the `reply` button in the `authwrapper`)

**Comment by thisislawatts** [comment]
@balhelhas and I had a talk through this today and explored a couple of different routes. Attached is a diagram that was used as an aid for that conversation. 

DataModelling

**Comment by thisislawatts** [comment]
@balhelhas thinking about the migration/rollout. What do you think about introducing a temp `useDiscussion` property which can be added to all new Howto/Research articles? This would mean we can introduce a simple branch to the FE for handling legacy vs. new implementation. 

Then we can use the absence of `useDiscussion` as an easy marker for documents whose comments need migrating.

**Comment by balhelhas** [comment]
@davehakkens @thisislawatts Hey guys, sorry for the delay on the PR. Wasn't able to give it my full attention the last couple of weeks, but it's finally here. As previously said we are going to iterate over this first phase where we only add the CRUD functionalities of the Discussion module (with nested comments). This implementation has a high reliance on recursiveness for the comments. For the howtos and updates that don't have a Discussion item we are forcing a creation of them when accessing the page.

**Comment by thisislawatts** [comment]
@balhelhas can you take a look at why the Lint code step is failing?

**Comment by balhelhas** [comment]
Hey @thisislawatts could you help on the circular dependency error?

File: packages/components/src/CommentItem/CommentItem.tsx [hunk 2]
```diff
@@ -6,6 +6,8 @@ import { EditComment } from '../EditComment/EditComment'
 import { LinkifyText } from '../LinkifyText/LinkifyText'
 import { Modal } from '../Modal/Modal'
 import { Username } from '../Username/Username'
+import { CreateComment } from '../CreateComment/CreateComment'
+import { CommentList } from '../CommentList/CommentList'
```
Comment (thisislawatts) [inline packages/components/src/CommentItem/CommentItem.tsx:10]: @balhelhas this is the source of the circular dependency.

**Comment by thisislawatts** [comment]
Closing the PR as this work is being continued in smaller PRS :) Such as #3100 and #3102. If anyone has questions on the state of this please ping me as the lead.

**Pull request closed without merging**